### DISCUSSION TITLE
Error when trying to check for row existence with AsyncSession

### DISCUSSION THREAD

**Discussion by gwerbin-tive** [sqlalchemy/sqlalchemy#10199] [Usage Questions] [answered]
Following the recommendation [here](https://github.com/sqlalchemy/sqlalchemy/discussions/9404#discussioncomment-5174411), I wrote the following function:

```python
async def check_if_exact_duplicate(
    db: AsyncSession,
    curr_rec_info: RecordInfo,
    curr_sens_data: EditableSensorData,
) -> bool:
    query = (
        select(Widget)
        .where(
            Widget.created_time == curr_rec_info.created_time,
            Widget.longitude.is_distinct_from(curr_sens_data.longitude),
            Widget.latitude.is_distinct_from(curr_sens_data.latitude),
            Widget.accuracy.is_distinct_from(curr_sens_data.accuracy),
        )
        .exists()
    )
    await db.scalar(query)
```

However I get the following exception:

```none
      File "/Users/.../src/myapp/api/v1/routes.py", line 312, in process_sensor_record
        is_exact_duplicate = await check_if_exact_duplicate(db, record_info, sensor_data)
      File "/Users/.../src/myapp/api/v1/routes.py", line 89, in check_if_exact_duplicate
        await db.scalar(query)
      File "/Users/.../.venv/lib/python3.10/site-packages/sqlalchemy/ext/asyncio/session.py", line 497, in scalar
        result = await greenlet_spawn(
      File "/Users/.../.venv/lib/python3.10/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 179, in greenlet_spawn
        result = context.switch(*args, **kwargs)
      File "/Users/.../.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2296, in scalar
        return self._execute_internal(
      File "/Users/.../.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2039, in _execute_internal
        statement = coercions.expect(roles.StatementRole, statement)
      File "/Users/.../.venv/lib/python3.10/site-packages/sqlalchemy/sql/coercions.py", line 442, in expect
        return impl._implicit_coercions(
      File "/Users/.../.venv/lib/python3.10/site-packages/sqlalchemy/sql/coercions.py", line 1201, in _implicit_coercions
        return super()._implicit_coercions(
      File "/Users/.../.venv/lib/python3.10/site-packages/sqlalchemy/sql/coercions.py", line 506, in _implicit_coercions
        self._raise_for_expected(element, argname, resolved)
      File "/Users/.../.venv/lib/python3.10/site-packages/sqlalchemy/sql/coercions.py", line 535, in _raise_for_expected
        raise exc.ArgumentError(msg, code=code) from err
    sqlalchemy.exc.ArgumentError: Executable SQL or text() construct expected, got <sqlalchemy.sql.selectable.Exists object at 0x1050d7190>.
```

Clearly I've misunderstood something in the linked post and/or the documentation. What is the correct technique for constructing an "existence" query that returns only `True` or `False`?

**Comment by zzzeek** [comment, marked as answer]
this is the same question as #9404 in fact.  the exists() must be **inside** of a `select()`:

    await db.scalar(select(exists_query))

an "exists" is a boolean column expression, like `True` or `False`.  It has to be in the form "SELECT EXISTS ...."

**Reply by gwerbin-tive** [reply]
Thanks, I completely misread the answer you gave in that question.

For posterity and for the sake of future readers who are as clumsy as I am, the full statement that worked was:

```python
async def check_if_exact_duplicate(
    db: AsyncSession,
    curr_rec_info: RecordInfo,
    curr_sens_data: EditableSensorData,
) -> bool:
    query = select(
        select(Widget)
        .where(
            Widget.created_time == curr_rec_info.created_time,
            Widget.longitude.is_distinct_from(curr_sens_data.longitude),
            Widget.latitude.is_distinct_from(curr_sens_data.latitude),
            Widget.accuracy.is_distinct_from(curr_sens_data.accuracy),
        )
        .exists()
    )
    await db.scalar(query)
```

The _inner_ `select(Widget).where(...).exists()` constructs the `exists` expression. The _outer_ `select()` then wraps it in a `SELECT` statement.

It seems like you could also use `exists().where(...)` for the inner `exists` expression, but the explicit version above seems easier to reason about for me.

Edit: Looking at the logs, `exists().where(...)` generates `EXISTS ( SELECT * ... )` while `select(Widget).where(...).exists()` generates `EXISTS ( SELECT col1, col2, ... )`. So I'll actually go with the `exists().where(...)` version.

**Reply by CaselIT** [reply]
you can also call `.select()` on an exists https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Exists.select

this way you avoid needed to call `select(the_exits)` and use the method chaining like the rest of the expresson

**Reply by gwerbin-tive** [reply]
> you can also call `.select()` on an exists https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Exists.select
> 
> this way you avoid needed to call `select(the_exits)` and use the method chaining like the rest of the expresson

Right @CaselIT. In my edit I noted that using a "top-level" `exists()` produced SQL closer to what I wanted, so that's what I went with in my own application.
